### DISCUSSION TITLE
Error with api.Select in a loop

### DISCUSSION THREAD

**Discussion by jeljelih** [Consensys/gnark#833] [Q&A] [answered]
I would like to call api.Select inside a for loop and I obtain the following error:
ERR parsing circuit error="assertIsBoolean failed: constant is not 0 or 1\nfrontend.parseCircuit.func2\n\tcompile.go:117\nr1cs.(*builder).AssertIsBoolean\n\tapi_assertions.go:52\nr1cs.(*builder).Select\n\tapi.go:440\nmain.(*MyCircuit).Define\n\tfirstCircuit.go:26\n"

This is my code:
`package main

import (
	"github.com/consensys/gnark-crypto/ecc"
	"github.com/consensys/gnark/backend/groth16"
	"github.com/consensys/gnark/frontend"
	"github.com/consensys/gnark/frontend/cs/r1cs"
)

// MyCircuit defines a simple circuit
type MyCircuit struct {
	// struct tags on a variable is optional
	// default uses variable name and secret visibility.
	Temp1 frontend.Variable
	Temp2 frontend.Variable
}

// Define declares the circuit constraints
func (circuit *MyCircuit) Define(api frontend.API) error {
	temp := api.Select(api.Cmp(3, circuit.Temp1), circuit.Temp2, 0)
	api.Println(temp)

	for i := 0; i < 10; i++ {
		temp1 := api.Select(api.Cmp(3, i), circuit.Temp2, 0)
		api.Println(temp1)
	}
	return nil
}

func main() {
	
	// compiles our circuit into a R1CS
	var circuit MyCircuit
	ccs, _ := frontend.Compile(ecc.BN254.ScalarField(), r1cs.NewBuilder, &circuit)

	// groth16 zkSNARK: Setup
	pk, vk, _ := groth16.Setup(ccs)

	// witness definition
	assignment := MyCircuit{
		Temp1: 2,
		Temp2: 35,
	}
	witness, _ := frontend.NewWitness(&assignment, ecc.BN254.ScalarField())
	publicWitness, _ := witness.Public()

	// groth16: Prove & Verify
	proof, _ := groth16.Prove(ccs, pk, witness)
	groth16.Verify(proof, vk, publicWitness)
}`

**Comment by ivokub** [comment, marked as answer]
`Cmp` returns either -1, 0 or 1 depending if x < y, x == y or x > y, but select assumes that input is either 0 or 1.

**Reply by jeljelih** [reply]
Thanks @ivokub for your prompt answer. I understand. Do you know how I can write a for loop that contains an if/else statement based on the value of the loop counter?

**Reply by ivokub** [reply]
In your current example you can split the loop into two: for iterations 0,1,2 and then for iterations 3,4,5,6,7,8,9:
```go
for i := 0; i < 3; i++ {
	temp1 := circuit.Temp2
	api.Println(temp1)
}

for i := 3; i < 10; i++ {
	temp1 := 0
	api.Println(temp1)
}
```

But it may be that I have misunderstood what you want to do.

**Reply by jeljelih** [reply]
Sorry I forgot to mention that the value 3 is an input to the circuit. 
This is the full implementation of what I want to do. Now,I obtain the error "unsatisfied constraint".

```
package main

import (
	"github.com/consensys/gnark-crypto/ecc"
	"github.com/consensys/gnark/backend/groth16"
	"github.com/consensys/gnark/frontend"
	"github.com/consensys/gnark/frontend/cs/r1cs"
)

// MyCircuit defines a simple circuit
type MyCircuit struct {
	// struct tags on a variable is optional
	// default uses variable name and secret visibility.
	Temp1 frontend.Variable
	Temp2 frontend.Variable
	Limit frontend.Variable
}

// Define declares the circuit constraints
func (circuit *MyCircuit) Define(api frontend.API) error {

	temp := api.Select(api.Cmp(5, circuit.Limit), circuit.Temp2, circuit.Temp1)
	api.Println(temp)
	api.AssertIsEqual(temp, circuit.Temp2)

	for i := 0; i < 10; i++ {
		api.Println(api.Cmp(i, circuit.Limit))
		temp := api.Select(api.Cmp(i, circuit.Limit), circuit.Temp2, circuit.Temp1)
		api.Println(temp)
	}
	return nil
}

func main() {
	
	// compiles our circuit into a R1CS
	var circuit MyCircuit
	ccs, _ := frontend.Compile(ecc.BN254.ScalarField(), r1cs.NewBuilder, &circuit)

	// groth16 zkSNARK: Setup
	pk, vk, _ := groth16.Setup(ccs)

	// witness definition
	assignment := MyCircuit{
		Temp1: 2,
		Temp2: 35,
		Limit: 3,
	}
	witness, _ := frontend.NewWitness(&assignment, ecc.BN254.ScalarField())
	publicWitness, _ := witness.Public()

	// groth16: Prove & Verify
	proof, _ := groth16.Prove(ccs, pk, witness)
	groth16.Verify(proof, vk, publicWitness)
}

```

**Reply by ivokub** [reply]
In the loop you have `api.Cmp(i, circuit.Limit)` which may be either -1, 0 or 1. But `api.Select` allows only inputs 0, 1. The behaviour for other cases (here, for the case -1) is undefined.

What behaviour are you seeking for? What should be chosen in the loop in cases?:
* `i < circuit.Limit`
* `i == circuit.Limit`
* `i > circuit.Limit`

**Reply by jeljelih** [reply]
Thanks for your reply @ivokub 
The behavior I am looking for is:
- for values below circuit.Limit, I want to select Temp1 value
- for values above circuit.Limit, I want to select Temp2 value

**Reply by ivokub** [reply]
And for value equaling `circuit.Limit`?

**Reply by jeljelih** [reply]
for value equaling `circuit.Limit`, I want to select Temp1 value

**Reply by ivokub** [reply]
Then it is going to be a bit more difficult. Essentially what you want to do is to have masked lists:
1) mask if `i < Limit`. Say this mask is `mask_Less`
2) mask if `i == Limit`. Sa this mask is `mask_isEq`
3) mask if `i > Limit`, call it `mask_More`

Then you set the value of the variable as `mask_less * Temp1 + mask_isEq * Temp1 + mask_more * Temp2`.

To compute the mask values, you can use `api.Cmp` and `api.IsZero` methods. Note that `api.Cmp` is expensive and you can optimise the computation of mask by using `std/math/cmp` gadget which provides bounded comparator implementation (as we know the limit on `i`.

**Reply by aybehrouz** [reply]
When binary output is needed, there is also some generic comparison [methods](https://pkg.go.dev/github.com/consensys/gnark@v0.9.0/std/math/cmp#IsLessOrEqual) in the `cmp` package. They may come in handy sometimes. As I remember the generated bit by bit comparison circuit is 1/3 smaller than `api.Cmp`.
